src/main/go/RelpConnection.go

```diff
@@ -0,0 +1,302 @@
+		go func() {
+			sendErr := relpConn.SendRelpRequest(relpRequest)
+			if sendErr != nil {
+				log.Printf("Error sending relp request: '%v'\n", err.Error())
+			}
+		}()
+	}
+
+	ackErr := relpConn.ReadAcks(batch)
+	if ackErr != nil {
+		// ACK timeout or other failure
+		return ackErr
+	}
+	return nil
+}
+
+// ReadAcks reads the ACKs from the given batch.
+func (relpConn *RelpConnection) ReadAcks(batch *RelpBatch) error {
+	log.Printf("ReadAcks.Entry>Reading ACKs for batchID: %v\n", batch.requestId)
+	var parser *RelpParser = nil
+	notComplete := relpConn.window.Size() > 0
+	var cn = *relpConn.connection
+
+	for notComplete {
+		tmp := make([]byte, 64)
+		readBytes := 0
+		for {
+			n, err := cn.Read(tmp)
+
+			err2 := cn.SetReadDeadline(time.Now().Add(relpConn.ackTimeoutDuration))
+			if err2 != nil {
+				return errors.New("error setting connection timeout")
+			}
+
+			if err != nil {
+				if errors.Is(err, os.ErrDeadlineExceeded) {
+					// reading timed out
+					return errors.New("reading ACK timed out")
+				} else if err == io.EOF {
+					return errors.New("encountered EOF in ACK (no connection?)")
+				} else {
+					// other error
+					return errors.New("unexpected error reading ACK")
+				}
+			}
+
+			// write and break if line break encountered
+			relpConn.preAllocRxBuffer.Write(tmp[0:n])
+			readBytes += n
+			if tmp[n-1] == '\n' {
+				break
+			}
+		}
+
+		parsedBytes := 0
+		if readBytes > 0 {
+			log.Printf("ReadAcks> Read %v byte(s) as ACK\n", readBytes)
+			for parsedBytes < readBytes {
+				if parser == nil {
+					parser = &RelpParser{}
+				}
+
+				nextBytes := relpConn.preAllocRxBuffer.Next(1) // len always 1
+				//log.Printf("Parsing byte: %v (str: %v)", nextBytes[0], string(nextBytes[0]))
+				parser.Parse(nextBytes[0])
+				parsedBytes++
+
+				if parser.isComplete {
+					log.Printf("ReadAcks> Parsing complete\n")
+					// resp read successfully
+					txnId := parser.frameTxnId
+					if relpConn.window.IsPending(txnId) {
+						reqId, err := relpConn.window.GetPending(txnId)
+						if err != nil {
+							panic("Could not find given pending txnId from RelpWindow!")
+						}
+						response := RelpFrameRX{
+							RelpFrame{
+								transactionId: parser.frameTxnId,
+								cmd:           parser.frameCmdString,
+								dataLength:    parser.frameLen,
+								data:          parser.frameData.Bytes(),
+							},
+						}
+						batch.PutResponse(reqId, &response)
+						relpConn.window.RemovePending(txnId)
+					}
+
+					parser = nil
+					if relpConn.window.Size() == 0 {
+						notComplete = false
+						break
+					}
+				}
+
+			}
+		}
+		// everything is read now
+		relpConn.preAllocRxBuffer.Reset()
+	}
+	log.Println("ReadAcks.Done> Return with no errors")
+	return nil
+}
+
+// SendRelpRequest sends the RELP frame to the connected RELP server
+func (relpConn *RelpConnection) SendRelpRequest(tx *RelpFrameTX) error {
+	buf := bytes.NewBuffer(make([]byte, 0, relpConn.txBufferSize))
+
+	txN, err := tx.Write(buf)
```

Comment:
is write timeout handled?

Reply:
write timeout handling added in [7a7367f](https://github.com/teragrep/rlp_05/pull/1/commits/7a7367ff670bc0b3e99dc638a43fa7ede9f0284b)